cmds/dump.c

```diff
@@ -1250,6 +1254,104 @@ static void dump_graphviz_footer(struct uftrace_dump_ops *ops, struct uftrace_da
 	graph_remove_task();
 }
 
+/* mermaid support */
+static struct uftrace_graph mermaid_graph = {
+	.root.head = LIST_HEAD_INIT(mermaid_graph.root.head),
+	.special_nodes = LIST_HEAD_INIT(mermaid_graph.special_nodes),
+};
+
+static void dump_mermaid_task_rstack(struct uftrace_dump_ops *ops, struct uftrace_task_reader *task,
+				     char *name)
+{
+	struct uftrace_record *frs = task->rstack;
+	struct uftrace_task_graph *graph;
+
+	graph = graph_get_task(task, sizeof(*graph));
+
+	graph->graph = &mermaid_graph;
+	mermaid_graph.sess = find_task_session(&task->h->sessions, task->t, frs->time);
+	if (graph->node == NULL)
+		graph->node = &mermaid_graph.root;
+
+	graph_add_node(graph, frs->type, name, sizeof(struct uftrace_graph_node), NULL);
+}
+
+static void dump_mermaid_kernel_rstack(struct uftrace_dump_ops *ops,
+				       struct uftrace_kernel_reader *kernel, int cpu,
+				       struct uftrace_record *rec, char *name)
+{
+	int tid;
+	struct uftrace_task_reader *task;
+	struct uftrace_task_graph *graph;
+
+	tid = kernel->tids[cpu];
+	task = get_task_handle(kernel->handle, tid);
+
+	graph = graph_get_task(task, sizeof(*graph));
+
+	graph->graph = &mermaid_graph;
+	mermaid_graph.sess = kernel->handle->sessions.first;
+
+	if (graph->node == NULL)
+		graph->node = &mermaid_graph.root;
+
+	graph_add_node(graph, rec->type, name, sizeof(struct uftrace_graph_node), NULL);
+}
+
+static void dump_mermaid_header(struct uftrace_dump_ops *ops, struct uftrace_data *handle,
+				struct uftrace_opts *opts)
+{
+	graph_init_callbacks(NULL, NULL, NULL, ops);
+	mermaid_graph.root.name = basename(handle->info.exename);
+
+	pr_out("<html>\n");
+	pr_out("<body>\n");
+	pr_out("<script src=\"https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js\"></script>\n");
+	pr_out("<script>\n");
+	pr_out("  var config = {\n");
+	pr_out("  startOnLoad: true,\n");
+	pr_out("  maxTextSize: 99999999,\n");
+	pr_out("  flowchart: { useMaxWidth: false }\n");
+	pr_out("  };\n");
+	pr_out("  mermaid.initialize(config);\n");
+	pr_out("</script>\n");
+}
+
+static void print_graph_node_mermaid(struct uftrace_graph *graph, struct uftrace_graph_node *node)
+{
+	char *symname = node->name;
+	struct uftrace_graph_node *child;
+
+	if (node->addr == EVENT_ID_PERF_SCHED_IN)
+		symname = "linux:schedule";
```

Comment:
I don't think we need this.  It should be handled by `graph_add_node()` already.

Reply:
I applied the code to delete this :)